libusb/os/windows_hotplug.c

```diff
@@ -0,0 +1,296 @@
+#include "libusbi.h"
+#include "threads_windows.h"
+
+#include "windows_common.h"
+#include "windows_hotplug.h"
+
+#include <stdio.h>
+#include <dbt.h>
+#include <usbiodef.h>
+
+static HWND windows_event_hWnd;
+static HANDLE windows_event_thread_handle;
+static DWORD WINAPI windows_event_thread_main(LPVOID lpParam);
+static LRESULT CALLBACK windows_proc_callback(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam);
+
+#define log_error(operation) do { \
+	usbi_err(NULL, "%s failed with error: %s", operation, windows_error_str(0)); \
+} while (0)
+
+int windows_start_event_monitor(void)
+{
+	windows_event_thread_handle = CreateThread(
+		NULL,
+		0,
+		windows_event_thread_main,
+		NULL,
+		0,
+		NULL
+	);
+
+	if(windows_event_thread_handle == NULL)
+	{
+		log_error("CreateThread");
+		return LIBUSB_ERROR_OTHER;
+	}
+
+	return LIBUSB_SUCCESS;
+}
+
+int windows_stop_event_monitor(void)
+{
+	if(windows_event_hWnd == NULL)
+	{
+		return LIBUSB_SUCCESS;
+	}
+
+	if(!SUCCEEDED(SendMessage(windows_event_hWnd, WM_CLOSE, 0, 0)))
+	{
+		log_error("SendMessage");
```

Comment:
this will print "windows_stop_event_monitor: SendMessage failed with error: [reason from windows]".

Reply:
So this is not really a failure (since not returned)? Should it be a libusb warning instead of error message?

Reply:
If we are printing an error, should ret also be changed?

Reply:
I reconsidered this and, well, in the unlikely case where SendMessage doesn't succeed here, we shall indeed not wait on the thread to terminate, because it won't terminate actually. In this case we better just return instead of being blocked here forever.

The same is (most probably) true for the WaitForSingleObject : it it fails, we shall not try to close the handle, it may result in unpredictable behaviors.

I think that if any of these fails, the event thread will be in an "unknown" state, but this is not supposed to be a problem when exiting the library, since the thread and its things will be destroyed anyway.

Reply:
Sounds good.